integrations/go/workload-executor.go

```diff
@@ -0,0 +1,315 @@
+		expectedID = int32(f)
+	}
+
+	return expectedID == nil || (actualResult != nil && expectedID == actualResult.InsertedID)
+}
+
+func verifyCursorResult(cur *mongo.Cursor, result interface{}) bool {
+	if result == nil {
+		return true
+	}
+
+	if cur == nil {
+		return false
+	}
+
+	defer func() {
+		cur.Close(context.Background())
+	}()
+
+	for _, expected := range result.(bson.A) {
+		if !cur.Next(context.Background()) {
+			return false
+		}
+		if !bytes.Equal(expected.(bson.Raw), cur.Current) {
+			return false
+		}
+	}
+
+	if cur.Next(context.Background()) {
+		return false
+	}
+	return cur.Err() == nil
+}
+
+func verifyUpdateResult(res *mongo.UpdateResult, result interface{}) bool {
+	if result == nil {
+		return true
+	}
+
+	var expected struct {
+		MatchedCount  int64 `bson:"matchedCount"`
+		ModifiedCount int64 `bson:"modifiedCount"`
+		UpsertedCount int64 `bson:"upsertedCount"`
+	}
+	err := bson.Unmarshal(result.(bson.Raw), &expected)
+	if err != nil {
+		return false
+	}
+
+	if expected.MatchedCount != res.MatchedCount {
+		return false
+	}
+	if expected.ModifiedCount != res.ModifiedCount {
+		return false
+	}
+
+	actualUpsertedCount := int64(0)
+	if res.UpsertedID != nil {
+		actualUpsertedCount = 1
+	}
+	return expected.UpsertedCount == actualUpsertedCount
+}
+
+func executeCollectionOperation(coll *mongo.Collection, op *operation) (bool, error) {
+	switch op.Name {
+	case "insertOne":
+		res, err := executeInsertOne(coll, op.Arguments)
+		return verifyInsertOneResult(res, op.Result), err
+	case "find":
+		cursor, err := executeFind(coll, op.Arguments)
+		return verifyCursorResult(cursor, op.Result), err
+	case "updateOne":
+		res, err := executeUpdateOne(coll, op.Arguments)
+		return verifyUpdateResult(res, op.Result), err
+	}
+	return false, errors.New("unrecognized collection operation: " + op.Name)
+}
+
+func runOperation(coll *mongo.Collection, op *operation) (bool, error) {
+	// execute the command on the given object
+	if op.Object == "collection" {
+		return executeCollectionOperation(coll, op)
+	}
+	return false, errors.New("unrecognized object: " + op.Name)
+}
+
+func main() {
+	connstring := os.Args[1]
+	workloadSpec := os.Args[2]
+
+	var workload driverWorkload
+
+	err := bson.UnmarshalExtJSONWithRegistry(specTestRegistry,[]byte(workloadSpec), false, &workload)
+	if err != nil {
+		str := "failed to unmarshal data: " + err.Error()
+		panic(str)
+	}
+
+	client, err := mongo.Connect(context.Background(), options.Client().ApplyURI(connstring))
+	if err != nil {
+		str := "failed to connect to client: " + err.Error()
+		panic(str)
+	}
+	defer func() { _ = client.Disconnect(context.Background()) }()
+
+	db := client.Database(workload.Database)
+	coll := db.Collection(workload.Collection)
+
+	results := struct {
+		NumErrors    int `json:"numErrors"`
+		NumFailures  int `json:"numFailures"`
+		NumSuccesses int `json:"numSuccesses"`
+	}{}
+
+	done := make(chan struct{})
+
+	go func() {
```

Comment:
Can you add a comment explaining what this goroutine is doing? Also, are we always guaranteed to get a SIGTERM?

Reply:
done. and yup, we repeat the operations until astrolabe sends a termination signal